contracts/Insurance.sol

```diff
@@ -137,29 +161,46 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
     function drainPool(uint256 amount) external override onlyLiquidation() {
         IERC20 tracerMarginToken = IERC20(tracer.tracerQuoteToken());
 
-        // Enforce a minimum. Very rare as funding rate will be incredibly high at this point
-        if (collateralAmount < 10**18) {
-            return;
-        }
+        uint256 poolHoldings = publicCollateralAmount + bufferCollateralAmount;
 
-        // Enforce a maximum at poolAmount
-        if (amount > collateralAmount) {
-            amount = collateralAmount;
-        }
+        if (amount >= poolHoldings) {
+            // If public collateral left after draining is less than 1 token, we want to keep it at 1 token
+            if (publicCollateralAmount > 10**18) {
+                // Leave 1 token for the public pool
+                amount = poolHoldings - 10**18;
+                publicCollateralAmount = 10**18;
+            } else {
+                amount = bufferCollateralAmount;
+            }
+
+            // Drain buffer
+            bufferCollateralAmount = 0;
+        } else if (amount > bufferCollateralAmount) {
+            if (publicCollateralAmount < 10**18) {
+                // If there's not enough public collateral for there to be 1 token, cap amount being drained at the buffer
+                amount = bufferCollateralAmount;
+            } else if (poolHoldings - amount < 10**18) {
+                // If the amount of collateral left in the public insurance would be less than 1 token, cap amount being drained
+                // from the public insurance such that 1 token is left in the public buffer
+                amount =
+                    bufferCollateralAmount +
+                    publicCollateralAmount - 
+                    10**18;
+                publicCollateralAmount = 10**18;
+            } else {
+                // Take out what you need from the public pool; there's enough for there to be >= 1 token left
+                publicCollateralAmount = poolHoldings - amount;
+            }
```

Comment:
this looks correct, but there's a lot going on here so I won't confirm - @BenjaminPatch to confirm

Reply:
There is a lot going on but there are a decent amount of edge cases that need checking. I'm confident that this is good.